core/src/main/java/com/sequenceiq/cloudbreak/service/publicendpoint/FreeIPAEndpointManagementService.java

```diff
@@ -0,0 +1,128 @@
+import com.sequenceiq.freeipa.api.v1.dns.model.AddDnsCnameRecordRequest;
+
+@Service
+public class FreeIPAEndpointManagementService extends BasePublicEndpointManagementService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIPAEndpointManagementService.class);
+
+    private static final String DOMAIN_PART_DELIMITER = ".";
+
+    @Inject
+    private DnsV1Endpoint dnsV1Endpoint;
+
+    @Inject
+    private EnvironmentClientService environmentClientService;
+
+    @Inject
+    private LoadBalancerPersistenceService loadBalancerPersistenceService;
+
+    @Inject
+    private LoadBalancerConfigService loadBalancerConfigService;
+
+    public void registerLoadBalancerDomainWithFreeIPA(Stack stack) {
+        Optional<LoadBalancer> loadBalancerOptional = getLoadBalancer(stack.getId());
+        if (loadBalancerOptional.isPresent()) {
+            LoadBalancer loadBalancer = loadBalancerOptional.get();
+            try {
+                if (StringUtils.isNotEmpty(loadBalancer.getDns())) {
+                    sendAddDnsCnameRecordRequest(stack, loadBalancer);
+                } else if (StringUtils.isNotEmpty(loadBalancer.getIp())) {
+                    sendAddDnsARecordRequest(stack, loadBalancer);
+                } else {
+                    LOGGER.error("Unable to find DNS or IP for load balancer. Load balancer will not be registered with FreeIPA.");
+                }
+            } catch (Exception e) {
+                LOGGER.error("Unable to register load balancer with FreeIPA", e);
+            }
+        } else {
+            LOGGER.debug("No load balancer found that needs to be registered with FreeIPA");
+        }
+    }
+
+    private void sendAddDnsCnameRecordRequest(Stack stack, LoadBalancer loadBalancer) {
+        String targetFQDN = StringUtils.appendIfMissing(loadBalancer.getDns(), DOMAIN_PART_DELIMITER);
+        String endpoint = loadBalancer.getEndpoint();
+
+        AddDnsCnameRecordRequest request = new AddDnsCnameRecordRequest();
+        request.setCname(endpoint);
+        request.setTargetFqdn(targetFQDN);
+        request.setEnvironmentCrn(stack.getEnvironmentCrn());
+        LOGGER.debug("Registering load balancer with target FQDN {} in FreeIPA with CNAME {}", targetFQDN, endpoint);
+        ThreadBasedUserCrnProvider.doAsInternalActor(() -> dnsV1Endpoint.addDnsCnameRecord(request));
+    }
+
+    private void sendAddDnsARecordRequest(Stack stack, LoadBalancer loadBalancer) {
+        String ip = loadBalancer.getIp();
+        String endpoint = loadBalancer.getEndpoint();
+
+        AddDnsARecordRequest request = new AddDnsARecordRequest();
+        request.setHostname(endpoint);
+        request.setIp(ip);
+        request.setEnvironmentCrn(stack.getEnvironmentCrn());
+        LOGGER.debug("Registering load balancer with target IP {} in FreeIPA with A record {}", ip, endpoint);
+        ThreadBasedUserCrnProvider.doAsInternalActor(() -> dnsV1Endpoint.addDnsARecord(request));
```

Comment:
this should call `addDnsARecordInternal`

Reply:
So should the full call look like:
```
ThreadBasedUserCrnProvider.doAsInternalActor(() -> dnsV1Endpoint.addDnsCnameRecordInternal(ThreadBasedUserCrnProvider.getAccountId(), request));
```

Reply:
yes, except you switched the method names. it should be for the A record

Reply:
Right, sorry, copy/pasted from the wrong line. Should be updated now.

Reply:
sorry, i haven't noticed it earlier, the call should look like this:
```
String accountId = ThreadBasedUserCrnProvider.getAccountId();
ThreadBasedUserCrnProvider.doAsInternalActor(() -> dnsV1Endpoint.addDnsARecordInternal(accountId, request));
```

Reply:
Done. I'll run another test to make sure it works before merging it in, but the code changes have been pushed for review.